state/machine.go

```diff
@@ -2237,40 +2239,56 @@ func (m *Machine) SetUpgradeCharmProfile(appName, chURL string) error {
 				return nil, errors.Trace(err)
 			}
 		}
+
 		var emptyProfile bool
 		if ch == nil || (ch.LXDProfile() == nil || ch.LXDProfile().Empty()) {
 			emptyProfile = true
 		}
 
-		provisioned := true
+		ops := []txn.Op{
+			{
+				C:      machinesC,
+				Id:     m.doc.DocID,
+				Assert: isAliveDoc,
+			},
+			// TODO: lxd profile - enhance checks for this set of transactions
+			//}, {
+			//	C:      charmsC,
+			//	Id:     ch.doc.DocID,
+			//	Assert: bson.D{{"url", ch.URL()}},
+			//},
+		}
+
 		profiles, err := m.CharmProfiles()
 		if err != nil {
 			if !errors.IsNotProvisioned(err) {
 				return nil, errors.Trace(err)
 			}
-			provisioned = false
-		}
-		// we need to ensure that the machine is provisioned before we attempt
-		// to set that the upgrade charm profile complete to not required.
-		// If the profiles are empty and there is nothing to apply, we can
-		// assume that nothing is required.
-		if len(profiles) == 0 && emptyProfile {
-			var ops []txn.Op
-			if provisioned {
-				ops = append(ops, m.checkCharmProfilesIsEmptyOp())
+		} else {
```

Comment:
can we get a comment here? I think using an explicit flag might be clearer, eg:

if err != nil {
  if errors.IsNotProvisioned(err) {
    // we don't need to fail here
    provisioned = false
  } else {
    return nil, errors.Trace(err)
  }

if provisioned {
  ops = append(ops, txn.Op{...}
}

Might more clearly convey why the code is structured that way.

Reply:
restructured the code section and added comments.